src/decode.rs

```diff
@@ -144,6 +178,74 @@ pub fn decode_into(input: &[u8], output: &mut [u8], alpha: &[u8; 58]) -> Result<
     Ok(index)
 }
 
+/// Decode given input slice into given output slice using the given alphabet. Expects
+/// and validates checksum bytes in input.
+///
+/// Option for version byte. If given, it is used to verify input.
+///
+/// Returns the length written into the byte slice, the rest of the bytes in
+/// the slice will be left untouched.
+///
+/// This is the low-level implementation that the `DecodeBuilder` uses to
+/// perform the decoding, it's very likely that the signature will change if
+/// the major version changes.
+///
+/// # Examples
+///
+/// ```rust
+/// let input = "PWEu9GGN";
+/// let mut output = [0; 6];
+/// let l = bs58::decode::decode_check_into(input.as_ref(), &mut output, bs58::alphabet::DEFAULT, None);
+/// assert_eq!([0x2d, 0x31], output[..l.unwrap()]);
+/// ```
+#[cfg(feature = "check")]
+pub fn decode_check_into(input: &[u8], output: &mut [u8], alpha: &[u8; 58], expected_ver: Option<u8>) -> Result<usize, DecodeError> {
+    use sha2::{Sha256, Digest};
+    use CHECKSUM_LEN;
+
+    let decoded_len = decode_into(input, output, alpha)?;
+    if decoded_len < CHECKSUM_LEN {
+        return Err(DecodeError::NoChecksum)
+    }
+    let checksum_index = decoded_len - CHECKSUM_LEN;
+
+    let expected_checksum = &output[checksum_index..decoded_len];
+
+    let first_hash = Sha256::digest(&output[0..checksum_index]);
+    let second_hash = Sha256::digest(&first_hash);
+    let (checksum, _) = second_hash.split_at(CHECKSUM_LEN);
+
+    match checksum == expected_checksum {
+        true => {
+            match expected_ver {
+                Some(ver) => {
+                    if output[0] == ver {
+                        Ok(checksum_index)
+                    }
+                    else {
+                        Err(DecodeError::InvalidVersion{ver: output[0], expected_ver: ver})
+                    }
+                }
+                None => Ok(checksum_index)
+            }
+
+        },
+        false => {
+            let mut a: [u8; CHECKSUM_LEN] = Default::default();
+            a.copy_from_slice(&checksum[..]);
+            let mut b: [u8; CHECKSUM_LEN] = Default::default();
+            b.copy_from_slice(&expected_checksum[..]);
+            Err(DecodeError::InvalidChecksum{checksum:a, expected_checksum:b})
+        }
+    }
+}
+
+
+#[cfg(not(feature = "check"))]
+pub fn decode_check_into(_input: &[u8], _output: &mut [u8], _alpha: &[u8; 58], _expected_ver: Option<u8>) -> Result<usize, DecodeError> {
```

Comment:
To avoid the "undocumented public function" error you could move the feature check onto a hidden inner function like

```rust
/// docs
pub fn decode_check_into(...) {
    #[cfg(feature = "check")]
    fn impl(...) {
    }

    #[cfg(not(feature = "check"))]
    fn impl(...) {
    }

    impl(...)
}
```

That way the public interface isn't affected by the feature, and the private functions are fully hidden behind it.

---

It would also be good to mention the panicking in the doc comment. Since this is pretty close to internal implementation details I'm fine with having it panic when the feature is turned off, but if anyone does use this directly they will want to know.

Reply:
trying this but it tries to run the doc code even when the feature is not included. I was thinking of just including a simple doc string saying it requires the feature. 

///// This function is used with check feature. A non-implemented function is included
///// when the feature is not used.

Reply:
Ah, yeah, it's possible to make doc-tests feature dependent, but not very nice, since doc-comments are attributes you can use `cfg_attr` to mark the test as ignored on some `cfg`:

```rust
/// # Examples
///
#[cfg_attr(feature = "check", doc("```"))]
#[cfg_attr(not(feature = "check"), doc("```ignore"))]
/// code()
/// ```
```

I'm happy either way, up to you if it's worth it.